batch_job/src/batch_job_mesos.c

```diff
@@ -0,0 +1,437 @@
+		}
+		fprintf(task_info_fp, "%s,", text_list_get(mt->task_input_files, num_input_files-1));
+
+	} else {
+		fputs(",", task_info_fp);
+	}
+    
+	if (extra_output_files != NULL && strlen(extra_output_files) != 0) {
+		int j = 0;
+		int num_output_files = text_list_size(mt->task_output_files);
+		for (j = 0; j < (num_output_files-1); j++) {
+			fprintf(task_info_fp, "%s ", text_list_get(mt->task_output_files, j));
+		}
+		fprintf(task_info_fp, "%s,",text_list_get(mt->task_output_files, num_output_files-1));
+	} else {
+		fputs(",", task_info_fp);
+	}
+
+	// The default resource requirements for each task
+	int64_t cores = -1;
+	int64_t memory = -1;
+	int64_t disk = -1;
+
+	if (resources) {
+		cores  = resources->cores  > -1 ? resources->cores  : cores;
+		memory = resources->memory > -1 ? resources->memory : memory;
+		disk   = resources->disk   > -1 ? resources->disk   : disk;
+	}
+
+	fprintf(task_info_fp, "%" PRId64 ",%" PRId64 ",%" PRId64 ",", cores, memory, disk);
+
+	fputs("submitted\n", task_info_fp);
+
+	mesos_task_delete(mt);
+	fclose(task_info_fp);
+
+	return task_id;
+}
+
+static batch_job_id_t batch_job_mesos_wait (struct batch_queue * q, struct batch_job_info * info_out, time_t stoptime)
+{
+		
+	char *line = NULL;
+	size_t len = 0;
+	ssize_t read_len;
+	FILE *task_state_fp;
+
+	if(finished_tasks == NULL) {
+		finished_tasks = itable_create(0);
+	}
+
+	while(access(FILE_TASK_STATE, F_OK) == -1) {}
+
+	task_state_fp = fopen(FILE_TASK_STATE, "r");
+
+	while(1) {
+
+		char *task_id_str;
+		char *task_stat_str;
+		const char *task_exit_code;
+		int task_id;
+				
+		while((read_len = getline(&line, &len, task_state_fp)) != -1) {
+
+			// trim the newline character
+			if (line[read_len-1] == '\n') {
+				line[read_len-1] = '\0';
+				--read_len;
+			}
+
+			task_id_str = strtok(line, ",");
+			task_id = atoi(task_id_str);
+
+			// There is a new task finished
+			if(itable_lookup(finished_tasks, task_id) == NULL) {
+
+				struct batch_job_info *info = itable_remove(q->job_table, task_id);
+			    	
+				info->finished = time(0);
+				task_stat_str = strtok(NULL, ",");
+
+				if (strcmp(task_stat_str, "finished") == 0) {
+					info->exited_normally = 1;
+				} else if (strcmp(task_stat_str, "failed") == 0) {
+					info->exited_normally = 0;
+					task_exit_code = strtok(NULL, ",");
+					if(atoi(task_exit_code) == 444) {
+						info->exit_code = 444;
+						debug(D_BATCH, "Task %s failed to retrieve the output.", task_id_str);
+					}
+					info->exit_code = atoi(task_exit_code);
+				} else {
+					info->exited_normally = 0;
+				}
+
+				memcpy(info_out, info, sizeof(*info));
+				free(info);
+				fclose(task_state_fp);
+
+				int itable_val = 1;
+				itable_insert(finished_tasks, task_id, &itable_val);
+
+				return task_id;
+			}
+		}
+		sleep(1);
+
+		if(stoptime != 0 && time(0) >= stoptime) {
+			fclose(task_state_fp);
+			return -1;
+		}
+	}
+
+}
+
+static int batch_job_mesos_remove (struct batch_queue *q, batch_job_id_t jobid)
```

Comment:
The timeout is for checking if mesos successfully stop the container in the cluster in 40 seconds. If the timeout is hit, then we will say the `batch_job_mesos_remove` failed. Since it is hard to say if the mesos can stop the container in 40 seconds, I just disabled the timeout.